File: solutions/haskell/sum-of-multiples/1/src/SumOfMultiples.hs
module SumOfMultiples (sumOfMultiples) where

import qualified Data.List as DList

maxMultiple :: Integer -> Integer -> Integer
maxMultiple limit num
  |  num == 0               = 0
  |  limit `mod` num == 0   = quotient - 1
  |  otherwise              = quotient
    where quotient = limit `div` num

sumOfMultiples :: [Integer] -> Integer -> Integer
sumOfMultiples factors limit = 
  (sum . DList.nub) [x*y | x <- factors, y <- [1..(maxMultiple limit x)]]
    

File: solutions/javascript/sum-of-multiples/1/sum-of-multiples.js
export const sum = (factors, limit) => {
  const multiples = factors.reduce((set, factor) => {
    if (!factor) return set;
    let multiple = factor;
    while (multiple < limit) {
      set.add(multiple);
      multiple += factor;
    }
    return set;
  }, new Set([0]));
  return multiples.values().reduce((acc, val) => acc + val);
};
